helix-core/src/test/java/org/apache/helix/controller/stages/TestRebalancePipeline.java

```diff
@@ -61,8 +62,9 @@ public void testDuplicateMsg() {
     HelixManager manager = new DummyClusterManager(clusterName, accessor);
     ClusterEvent event = new ClusterEvent(ClusterEventType.Unknown);
     event.addAttribute(AttributeName.helixmanager.name(), manager);
-    event.addAttribute(AttributeName.ControllerDataProvider.name(),
-        new ResourceControllerDataProvider());
+    ResourceControllerDataProvider dataCache = new ResourceControllerDataProvider();
+    dataCache.setAsyncTasksThreadPool(Executors.newSingleThreadExecutor());
```

Comment:
Could you add a comment here stating why we need to set async tasks thread pool? What happens if we don't?

Reply:
Sure. I'll add. The reason I found this issue is that when I changed the session Id to be a valid one, the test actually failed. And debugging, I found it's due to the Executor was not set up, and thus the periodically clean up job for pending messages throw an NPE, which caused the pipeline to stop.

Reply:
I see. Although not directly related to this PR, do you think we should consider handling this NPE? It seems like something we should handle more gracefully (we probably won't have this issue in production).

Reply:
Our controller has properly set this field during construction: cache.setAsyncTasksThreadPool(_asyncTasksThreadPool);
I'd think the code itself should be safe.